vespalib/src/vespa/vespalib/stllike/cache.hpp

```diff
@@ -230,6 +253,64 @@ cache<P>::cache(BackingStore& backing_store, size_t max_bytes)
     : cache(backing_store, max_bytes, 0)
 {}
 
+template<typename P>
+void cache<P>::set_frequency_sketch_size(size_t cache_max_elem_count) {
+    std::lock_guard guard(_hashLock);
+    if (cache_max_elem_count > 0) {
+        _sketch = std::make_unique<SketchType>(cache_max_elem_count, _hasher);
+        // (Re)setting the sketch loses all frequency knowledge, but we can at the
+        // very least pre-seed it with the information we _do_ have, which is that
+        // all elements already in the cache have an estimated frequency of >= 1.
+        auto pre_seed_sketch = [this](const K& key) { _sketch->add(key); };
+        _probationary_segment.for_each_key(pre_seed_sketch);
+        _protected_segment.for_each_key(pre_seed_sketch); // no-op unless SLRU
+    } else {
+        _sketch.reset();
+    }
+}
+
+template<typename P>
+void
+cache<P>::lfu_add(const K& key) noexcept {
+    if (_sketch) {
+        _sketch->add(key);
+    }
+}
+
+template<typename P>
+uint8_t
+cache<P>::lfu_add_and_count(const K& key) noexcept {
+    return _sketch ? _sketch->add_and_count(key) : 0;
+}
+
+template<typename P>
+bool
+cache<P>::lfu_accepts_insertion(const K& key, const V& value,
+                                const SizeConstrainedLru& segment,
+                                uint8_t candidate_freq) const noexcept
+{
+    if (!_sketch) {
+        return true; // Trivially accepts insertion, since there's no LFU policy
+    }
+    // TODO > capacity_bytes() instead of >=, this uses >= to be symmetric with removeOldest()
+    const bool would_displace = ((segment.size() >= segment.capacity()) ||
+                                 (segment.size_bytes() + calcSize(key, value)) >= segment.capacity_bytes());
+    const K* victim;
+    if (would_displace && (victim = segment.last_key_or_nullptr()) != nullptr) {
```

Comment:
consider re-writing to not use assignment inside check

Reply:
Refactored to instead early-out on `!would_displace`